bigframes/core/nodes.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+import typing
+from typing import Optional, Tuple
+
+import pandas
+
+import bigframes.core.guid
+from bigframes.core.ordering import OrderingColumnReference
+import bigframes.core.window_spec as window
+import bigframes.dtypes
+import bigframes.operations as ops
+import bigframes.operations.aggregations as agg_ops
+
+if typing.TYPE_CHECKING:
+    import ibis.expr.types as ibis_types
+
+    import bigframes.core.ordering as orderings
+    import bigframes.session
+
+
+@dataclass(frozen=True)
+class BigFrameNode:
+    pass
+
+    @property
+    def deterministic(self) -> bool:
+        return True
+
+    @property
+    def child_nodes(self) -> typing.Sequence[BigFrameNode]:
+        return tuple([])
+
+    @property
+    def sessions(self):
```

Comment:
This walks the whole tree. I'd like something that can short-circuit that.

Reply:
Could we get away with something like "first_session" since these should all be the same object ideally.

Reply:
made it max 1 session and throw error if >1